File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-business-events-5-4/CompressionActivity.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class CompressionActivity : ISerialized
    {
        public CompressionActivity(XmlNode xmlNode)
        {
            XmlNode compressionTypeNode = xmlNode.SelectSingleNode("compressionType");
            
            if (compressionTypeNode != null)
            {
                if (compressionTypeNode.Attributes["href"] != null || compressionTypeNode.Attributes["id"] != null) 
                {
                    if (compressionTypeNode.Attributes["id"] != null) 
                    {
                        compressionTypeIDRef_ = compressionTypeNode.Attributes["id"].Value;
                        CompressionType ob = new CompressionType(compressionTypeNode);
                        IDManager.SetID(compressionTypeIDRef_, ob);
                    }
                    else if (compressionTypeNode.Attributes["href"] != null)
                    {
                        compressionTypeIDRef_ = compressionTypeNode.Attributes["href"].Value;
                    }
                    else
                    {
                        compressionType_ = new CompressionType(compressionTypeNode);
                    }
                }
                else
                {
                    compressionType_ = new CompressionType(compressionTypeNode);
                }
            }
            
        
            XmlNode replacementTradeIdentifierNode = xmlNode.SelectSingleNode("replacementTradeIdentifier");
            
            if (replacementTradeIdentifierNode != null)
            {
                if (replacementTradeIdentifierNode.Attributes["href"] != null || replacementTradeIdentifierNode.Attributes["id"] != null) 
                {
                    if (replacementTradeIdentifierNode.Attributes["id"] != null) 
                    {
                        replacementTradeIdentifierIDRef_ = replacementTradeIdentifierNode.Attributes["id"].Value;
                        TradeIdentifier ob = new TradeIdentifier(replacementTradeIdentifierNode);
                        IDManager.SetID(replacementTradeIdentifierIDRef_, ob);
                    }
                    else if (replacementTradeIdentifierNode.Attributes["href"] != null)
                    {
                        replacementTradeIdentifierIDRef_ = replacementTradeIdentifierNode.Attributes["href"].Value;
                    }
                    else
                    {
                        replacementTradeIdentifier_ = new TradeIdentifier(replacementTradeIdentifierNode);
                    }
                }
                else
                {
                    replacementTradeIdentifier_ = new TradeIdentifier(replacementTradeIdentifierNode);
                }
            }
            
        
            XmlNodeList originatingTradeIdentifierNodeList = xmlNode.SelectNodes("originatingTradeIdentifier");
            
            if (originatingTradeIdentifierNodeList != null)
            {
                this.originatingTradeIdentifier_ = new List<TradeIdentifier>();
                foreach (XmlNode item in originatingTradeIdentifierNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            originatingTradeIdentifierIDRef_ = item.Attributes["id"].Value;
                            originatingTradeIdentifier_.Add(new TradeIdentifier(item));
                            IDManager.SetID(originatingTradeIdentifierIDRef_, originatingTradeIdentifier_[originatingTradeIdentifier_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            originatingTradeIdentifierIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        originatingTradeIdentifier_.Add(new TradeIdentifier(item));
                        }
                    }
                    else
                    {
                        originatingTradeIdentifier_.Add(new TradeIdentifier(item));
                    }
                }
            }
            
        
            XmlNode replacementTradeIdNode = xmlNode.SelectSingleNode("replacementTradeId");
            
            if (replacementTradeIdNode != null)
            {
                if (replacementTradeIdNode.Attributes["href"] != null || replacementTradeIdNode.Attributes["id"] != null) 
                {
                    if (replacementTradeIdNode.Attributes["id"] != null) 
                    {
                        replacementTradeIdIDRef_ = replacementTradeIdNode.Attributes["id"].Value;
                        TradeId ob = new TradeId(replacementTradeIdNode);
                        IDManager.SetID(replacementTradeIdIDRef_, ob);
                    }
                    else if (replacementTradeIdNode.Attributes["href"] != null)
                    {
                        replacementTradeIdIDRef_ = replacementTradeIdNode.Attributes["href"].Value;
                    }
                    else
                    {
                        replacementTradeId_ = new TradeId(replacementTradeIdNode);
                    }
                }
                else
                {
                    replacementTradeId_ = new TradeId(replacementTradeIdNode);
                }
            }
            
        
            XmlNodeList originatingTradeIdNodeList = xmlNode.SelectNodes("originatingTradeId");
            
            if (originatingTradeIdNodeList != null)
            {
                this.originatingTradeId_ = new List<TradeId>();
                foreach (XmlNode item in originatingTradeIdNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            originatingTradeIdIDRef_ = item.Attributes["id"].Value;
                            originatingTradeId_.Add(new TradeId(item));
                            IDManager.SetID(originatingTradeIdIDRef_, originatingTradeId_[originatingTradeId_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            originatingTradeIdIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        originatingTradeId_.Add(new TradeId(item));
                        }
                    }
                    else
                    {
                        originatingTradeId_.Add(new TradeId(item));
                    }
                }
            }
            
        
        }
        
    
        #region CompressionType_
        private CompressionType compressionType_;
        public CompressionType CompressionType_
        {
            get
            {
                if (this.compressionType_ != null)
                {
                    return this.compressionType_; 
                }
                else if (this.compressionTypeIDRef_ != null)
                {
                    compressionType_ = IDManager.getID(compressionTypeIDRef_) as CompressionType;
                    return this.compressionType_; 
                }
                else
                {
                      return this.compressionType_; 
                }
            }
            set
            {
                if (this.compressionType_ != value)
                {
                    this.compressionType_ = value;
                }
            }
        }
        #endregion
        
        public string compressionTypeIDRef_ { get; set; }
        #region ReplacementTradeIdentifier_
        private TradeIdentifier replacementTradeIdentifier_;
        public TradeIdentifier ReplacementTradeIdentifier_
        {
            get
            {
                if (this.replacementTradeIdentifier_ != null)
                {
                    return this.replacementTradeIdentifier_; 
                }
                else if (this.replacementTradeIdentifierIDRef_ != null)
                {
                    replacementTradeIdentifier_ = IDManager.getID(replacementTradeIdentifierIDRef_) as TradeIdentifier;
                    return this.replacementTradeIdentifier_; 
                }
                else
                {
                      return this.replacementTradeIdentifier_; 
                }
            }
            set
            {
                if (this.replacementTradeIdentifier_ != value)
                {
                    this.replacementTradeIdentifier_ = value;
                }
            }
        }
        #endregion
        
        public string replacementTradeIdentifierIDRef_ { get; set; }
        #region OriginatingTradeIdentifier_
        private List<TradeIdentifier> originatingTradeIdentifier_;
        public List<TradeIdentifier> OriginatingTradeIdentifier_
        {
            get
            {
                if (this.originatingTradeIdentifier_ != null)
                {
                    return this.originatingTradeIdentifier_; 
                }
                else if (this.originatingTradeIdentifierIDRef_ != null)
                {
                    return this.originatingTradeIdentifier_; 
                }
                else
                {
                      return this.originatingTradeIdentifier_; 
                }
            }
            set
            {
                if (this.originatingTradeIdentifier_ != value)
                {
                    this.originatingTradeIdentifier_ = value;
                }
            }
        }
        #endregion
        
        public string originatingTradeIdentifierIDRef_ { get; set; }
        #region ReplacementTradeId_
        private TradeId replacementTradeId_;
        public TradeId ReplacementTradeId_
        {
            get
            {
                if (this.replacementTradeId_ != null)
                {
                    return this.replacementTradeId_; 
                }
                else if (this.replacementTradeIdIDRef_ != null)
                {
                    replacementTradeId_ = IDManager.getID(replacementTradeIdIDRef_) as TradeId;
                    return this.replacementTradeId_; 
                }
                else
                {
                      return this.replacementTradeId_; 
                }
            }
            set
            {
                if (this.replacementTradeId_ != value)
                {
                    this.replacementTradeId_ = value;
                }
            }
        }
        #endregion
        
        public string replacementTradeIdIDRef_ { get; set; }
        #region OriginatingTradeId_
        private List<TradeId> originatingTradeId_;
        public List<TradeId> OriginatingTradeId_
        {
            get
            {
                if (this.originatingTradeId_ != null)
                {
                    return this.originatingTradeId_; 
                }
                else if (this.originatingTradeIdIDRef_ != null)
                {
                    return this.originatingTradeId_; 
                }
                else
                {
                      return this.originatingTradeId_; 
                }
            }
            set
            {
                if (this.originatingTradeId_ != value)
                {
                    this.originatingTradeId_ = value;
                }
            }
        }
        #endregion
        
        public string originatingTradeIdIDRef_ { get; set; }
        public string choiceStr_0;
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-business-events-5-4/CompressionActivity.cpp
// CompressionActivity.cpp 
#include "CompressionActivity.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

CompressionActivity::CompressionActivity(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //compressionTypeNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* compressionTypeNode = xmlNode->FirstChildElement("compressionType");

   if(compressionTypeNode){compressionTypeIsNull_ = false;}
   else{compressionTypeIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- compressionTypeNode , address : " << compressionTypeNode << std::endl;
   #endif
   if(compressionTypeNode)
   {
      if(compressionTypeNode->Attribute("href") || compressionTypeNode->Attribute("id"))
      {
          if(compressionTypeNode->Attribute("id"))
          {
             compressionTypeIDRef_ = compressionTypeNode->Attribute("id");
             compressionType_ = boost::shared_ptr<CompressionType>(new CompressionType(compressionTypeNode));
             compressionType_->setID(compressionTypeIDRef_);
             IDManager::instance().setID(compressionTypeIDRef_,compressionType_);
          }
          else if(compressionTypeNode->Attribute("href")) { compressionTypeIDRef_ = compressionTypeNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { compressionType_ = boost::shared_ptr<CompressionType>(new CompressionType(compressionTypeNode));}
   }

   //replacementTradeIdentifierNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* replacementTradeIdentifierNode = xmlNode->FirstChildElement("replacementTradeIdentifier");

   if(replacementTradeIdentifierNode){replacementTradeIdentifierIsNull_ = false;}
   else{replacementTradeIdentifierIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- replacementTradeIdentifierNode , address : " << replacementTradeIdentifierNode << std::endl;
   #endif
   if(replacementTradeIdentifierNode)
   {
      if(replacementTradeIdentifierNode->Attribute("href") || replacementTradeIdentifierNode->Attribute("id"))
      {
          if(replacementTradeIdentifierNode->Attribute("id"))
          {
             replacementTradeIdentifierIDRef_ = replacementTradeIdentifierNode->Attribute("id");
             replacementTradeIdentifier_ = boost::shared_ptr<TradeIdentifier>(new TradeIdentifier(replacementTradeIdentifierNode));
             replacementTradeIdentifier_->setID(replacementTradeIdentifierIDRef_);
             IDManager::instance().setID(replacementTradeIdentifierIDRef_,replacementTradeIdentifier_);
          }
          else if(replacementTradeIdentifierNode->Attribute("href")) { replacementTradeIdentifierIDRef_ = replacementTradeIdentifierNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { replacementTradeIdentifier_ = boost::shared_ptr<TradeIdentifier>(new TradeIdentifier(replacementTradeIdentifierNode));}
   }

   //originatingTradeIdentifierNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* originatingTradeIdentifierNode = xmlNode->FirstChildElement("originatingTradeIdentifier");

   if(originatingTradeIdentifierNode){originatingTradeIdentifierIsNull_ = false;}
   else{originatingTradeIdentifierIsNull_ = true;}

   if(originatingTradeIdentifierNode)
   {
      for(originatingTradeIdentifierNode; originatingTradeIdentifierNode; originatingTradeIdentifierNode = originatingTradeIdentifierNode->NextSiblingElement("originatingTradeIdentifier")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- originatingTradeIdentifierNode , address : " << originatingTradeIdentifierNode << std::endl;
          #endif
          if(originatingTradeIdentifierNode->Attribute("href") || originatingTradeIdentifierNode->Attribute("id"))
          {
              if(originatingTradeIdentifierNode->Attribute("id"))
              {
                  originatingTradeIdentifierIDRef_ = originatingTradeIdentifierNode->Attribute("id");
                  originatingTradeIdentifier_.push_back(boost::shared_ptr<TradeIdentifier>(new TradeIdentifier(originatingTradeIdentifierNode)));
                  originatingTradeIdentifier_.back()->setID(originatingTradeIdentifierIDRef_);
                  IDManager::instance().setID(originatingTradeIdentifierIDRef_, originatingTradeIdentifier_.back());
              }
              else if(originatingTradeIdentifierNode->Attribute("href")) { originatingTradeIdentifierIDRef_ = originatingTradeIdentifierNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { originatingTradeIdentifier_.push_back(boost::shared_ptr<TradeIdentifier>(new TradeIdentifier(originatingTradeIdentifierNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- originatingTradeIdentifierNode , address : " << originatingTradeIdentifierNode << std::endl;
       #endif
   }

   //replacementTradeIdNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* replacementTradeIdNode = xmlNode->FirstChildElement("replacementTradeId");

   if(replacementTradeIdNode){replacementTradeIdIsNull_ = false;}
   else{replacementTradeIdIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- replacementTradeIdNode , address : " << replacementTradeIdNode << std::endl;
   #endif
   if(replacementTradeIdNode)
   {
      if(replacementTradeIdNode->Attribute("href") || replacementTradeIdNode->Attribute("id"))
      {
          if(replacementTradeIdNode->Attribute("id"))
          {
             replacementTradeIdIDRef_ = replacementTradeIdNode->Attribute("id");
             replacementTradeId_ = boost::shared_ptr<TradeId>(new TradeId(replacementTradeIdNode));
             replacementTradeId_->setID(replacementTradeIdIDRef_);
             IDManager::instance().setID(replacementTradeIdIDRef_,replacementTradeId_);
          }
          else if(replacementTradeIdNode->Attribute("href")) { replacementTradeIdIDRef_ = replacementTradeIdNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { replacementTradeId_ = boost::shared_ptr<TradeId>(new TradeId(replacementTradeIdNode));}
   }

   //originatingTradeIdNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* originatingTradeIdNode = xmlNode->FirstChildElement("originatingTradeId");

   if(originatingTradeIdNode){originatingTradeIdIsNull_ = false;}
   else{originatingTradeIdIsNull_ = true;}

   if(originatingTradeIdNode)
   {
      for(originatingTradeIdNode; originatingTradeIdNode; originatingTradeIdNode = originatingTradeIdNode->NextSiblingElement("originatingTradeId")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- originatingTradeIdNode , address : " << originatingTradeIdNode << std::endl;
          #endif
          if(originatingTradeIdNode->Attribute("href") || originatingTradeIdNode->Attribute("id"))
          {
              if(originatingTradeIdNode->Attribute("id"))
              {
                  originatingTradeIdIDRef_ = originatingTradeIdNode->Attribute("id");
                  originatingTradeId_.push_back(boost::shared_ptr<TradeId>(new TradeId(originatingTradeIdNode)));
                  originatingTradeId_.back()->setID(originatingTradeIdIDRef_);
                  IDManager::instance().setID(originatingTradeIdIDRef_, originatingTradeId_.back());
              }
              else if(originatingTradeIdNode->Attribute("href")) { originatingTradeIdIDRef_ = originatingTradeIdNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { originatingTradeId_.push_back(boost::shared_ptr<TradeId>(new TradeId(originatingTradeIdNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- originatingTradeIdNode , address : " << originatingTradeIdNode << std::endl;
       #endif
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<CompressionType> CompressionActivity::getCompressionType()
{
   if(!this->compressionTypeIsNull_){
        if(compressionTypeIDRef_ != ""){
             return boost::shared_static_cast<CompressionType>(IDManager::instance().getID(compressionTypeIDRef_));
        }else{
             return this->compressionType_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<CompressionType>();
   }
}
boost::shared_ptr<TradeIdentifier> CompressionActivity::getReplacementTradeIdentifier()
{
   if(!this->replacementTradeIdentifierIsNull_){
        if(replacementTradeIdentifierIDRef_ != ""){
             return boost::shared_static_cast<TradeIdentifier>(IDManager::instance().getID(replacementTradeIdentifierIDRef_));
        }else{
             return this->replacementTradeIdentifier_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<TradeIdentifier>();
   }
}
std::vector<boost::shared_ptr<TradeIdentifier>> CompressionActivity::getOriginatingTradeIdentifier()
{
   if(!this->originatingTradeIdentifierIsNull_){
        if(originatingTradeIdentifierIDRef_ != ""){
             return this->originatingTradeIdentifier_;
        }else{
             return this->originatingTradeIdentifier_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<TradeIdentifier>>();
   }
}
boost::shared_ptr<TradeId> CompressionActivity::getReplacementTradeId()
{
   if(!this->replacementTradeIdIsNull_){
        if(replacementTradeIdIDRef_ != ""){
             return boost::shared_static_cast<TradeId>(IDManager::instance().getID(replacementTradeIdIDRef_));
        }else{
             return this->replacementTradeId_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<TradeId>();
   }
}
std::vector<boost::shared_ptr<TradeId>> CompressionActivity::getOriginatingTradeId()
{
   if(!this->originatingTradeIdIsNull_){
        if(originatingTradeIdIDRef_ != ""){
             return this->originatingTradeId_;
        }else{
             return this->originatingTradeId_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<TradeId>>();
   }
}
}

